src/libprojectM/Renderer/MilkdropWaveform.cpp

```diff
@@ -4,393 +4,613 @@
+            float dx = cosf(angle);
+            float dy = sinf(angle);
+
+            float edgeX[2]{
+                x * cosf(angle + PI * 0.5f) - dx * 3.0f,
+                x * cosf(angle + PI * 0.5f) + dx * 3.0f
+            };
+
+            float edgeY[2]{
+                x * sinf(angle + PI * 0.5f) - dy * 3.0f,
+                x * sinf(angle + PI * 0.5f) + dy * 3.0f
+            };
+
+            for (int i = 0; i < 2; i++)
+            {
+                for (int j = 0; j < 4; j++)
+                {
+                    float t;
+                    bool clip{ false };
+
+                    switch (j)
+                    {
+                        case 0:
+                            if (edgeX[i] > 1.1f)
+                            {
+                                t = (1.1f - edgeX[1 - i]) / (edgeX[i] - edgeX[1 - i]);
+                                clip = true;
+                            }
+                            break;
+
+                        case 1:
+                            if (edgeX[i] < -1.1f)
+                            {
+                                t = (-1.1f - edgeX[1 - i]) / (edgeX[i] - edgeX[1 - i]);
+                                clip = true;
+                            }
+                            break;
+
+                        case 2:
+                            if (edgeY[i] > 1.1f)
+                            {
+                                t = (1.1f - edgeY[1 - i]) / (edgeY[i] - edgeY[1 - i]);
+                                clip = true;
+                            }
+                            break;
+
+                        case 3:
+                            if (edgeY[i] < -1.1f)
+                            {
+                                t = (-1.1f - edgeY[1 - i]) / (edgeY[i] - edgeY[1 - i]);
+                                clip = true;
+                            }
+                            break;
+
+                        default:
+                            break;
+                    }
+
+                    if (clip)
+                    {
+                        float diffX = edgeX[i] - edgeX[1 - i];
+                        float diffY = edgeY[i] - edgeY[1 - i];
+                        edgeX[i] = edgeX[1 - i] + diffX * t;
+                        edgeY[i] = edgeY[1 - i] + diffY * t;
+                    }
+                }
+            }
+
+            dx = (edgeX[1] - edgeX[0]) / static_cast<float>(samples);
+            dy = (edgeY[1] - edgeY[0]) / static_cast<float>(samples);
+
+            float angle2 = atan2f(dy, dx);
+            float perpetualDX = cosf(angle2 + PI * 0.5f);
+            float perpetualDY = sinf(angle2 + PI * 0.5f);
+
+            if (mode == MilkdropWaveformMode::Line)
+            {
+                for (int i = 0; i < samples; i++)
+                {
+                    wave1Vertices[i].x =
+                        edgeX[0] + dx * static_cast<float>(i) + perpetualDX * 0.25f * pcmDataL[i + sampleOffset];
```

Comment:
Just curious why we only use pcm left data here?

Reply:
It's the "single line" waveform mode. Milkdrop also only uses the left channel here, so I just did the same for maximum compatibility.

Original line of code is here, note the `fL[i + sample_offset]`, which is the equivalent of `pcmData[]`: https://github.com/projectM-visualizer/milkdrop2/blob/f05b0d811a87a17c4624170c26c93bac39b05bde/src/vis_milk2/milkdropfs.cpp#L3206